gxtest/cmpr.cpp

```diff
@@ -0,0 +1,299 @@
+  int red1 = Convert5To8((c1 >> 11) & 0x1F);
+  int red2 = Convert5To8((c2 >> 11) & 0x1F);
+
+  u16 ss = s & 3;
+  u16 tt = t & 3;
+
+  int colorSel = dxtBlock->lines[tt];
+  int rs = 6 - (ss << 1);
+  colorSel = (colorSel >> rs) & 3;
+  colorSel |= c1 > c2 ? 0 : 4;
+
+  struct Color color = {0, 0, 0, 0};
+
+  switch (colorSel) {
+  case 0:
+  case 4:
+    color = MakeRGBA(red1, green1, blue1, 255);
+    break;
+  case 1:
+  case 5:
+    color = MakeRGBA(red2, green2, blue2, 255);
+    break;
+  case 2:
+    color = MakeRGBA(DXTBlend(red2, red1), DXTBlend(green2, green1),
+                     DXTBlend(blue2, blue1), 255);
+    break;
+  case 3:
+    color = MakeRGBA(DXTBlend(red1, red2), DXTBlend(green1, green2),
+                     DXTBlend(blue1, blue2), 255);
+    break;
+  case 6:
+    color = MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2,
+                     (blue1 + blue2) / 2, 255);
+    break;
+  case 7:
+    color = MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2,
+                     (blue1 + blue2) / 2, 0);
+    break;
+  }
+  return color;
+}
+
+static void CmprTest() {
+  START_TEST();
+
+  // setup the fifo and then init the flipper
+  void *gp_fifo = memalign(32, DEFAULT_FIFO_SIZE);
+
+  memset(gp_fifo, 0, DEFAULT_FIFO_SIZE);
+  if (!gp_fifo)
+    return;
+
+  GX_Init(gp_fifo, DEFAULT_FIFO_SIZE);
+
+  u8 *tmp = (u8 *)memalign(32, OUT_SIZE); // this is probably the wrong size
+  u8 *out = (u8 *)memalign(32, OUT_SIZE);
+  if (!out)
+    return;
+  memset(out, 0, OUT_SIZE);
+  DCFlushRange(out, OUT_SIZE);
+
+  // clears the bg to color and clears the z buffer
+  GXColor background = {1, 0, 0, 0xff};
+  GX_SetCopyClear(background, 0x00ffffff);
+
+  GX_CopyDisp(tmp, GX_TRUE);
+
+  // setup the copy for later
+  GX_SetDispCopyYScale(1);
+  GX_SetCopyFilter(GX_FALSE, 0, GX_FALSE, 0);
+  GX_SetTexCopySrc(0, 0, OUT_WIDTH, OUT_HEIGHT);
+  GX_SetTexCopyDst(OUT_WIDTH, OUT_HEIGHT, GX_TF_RGBA8, 0);
+
+  {
+    // setup some other defaults
+    GX_SetCullMode(GX_CULL_NONE);
+    GX_SetViewport(0, 0, EFB_WIDTH, EFB_HEIGHT, 0, 1);
+    GX_SetScissor(0, 0, EFB_WIDTH, EFB_HEIGHT);
+    GX_SetVtxAttrFmt(GX_VTXFMT0, GX_VA_POS, GX_POS_XY, GX_F32, 0);
+    GX_SetVtxAttrFmt(GX_VTXFMT0, GX_VA_TEX0, GX_TEX_ST, GX_F32, 0);
+
+    GX_SetNumChans(1);
+    GX_SetNumTexGens(1);
+    GX_SetTevOp(GX_TEVSTAGE0, GX_REPLACE);
+    GX_SetTevOrder(GX_TEVSTAGE0, GX_TEXCOORD0, GX_TEXMAP0, GX_COLOR0A0);
+    GX_SetTexCoordGen(GX_TEXCOORD0, GX_TG_MTX2x4, GX_TG_TEX0, GX_IDENTITY);
+  }
+
+  // clear buffer
+  GX_CopyTex(out, GX_TRUE);
+
+  u8 *texdata;
+  {
+#define TX_WIDTH 512
+#define TX_HEIGHT 512
+    const u32 width = TX_WIDTH;
+    const u32 height = TX_HEIGHT;
+    const u32 bytesize = (width / 4) * (height / 4) * 8;
+    srand(0);
+    texdata = (u8 *)memalign(32, bytesize);
+    {
+      u32 i;
+      for (i = 0; i < bytesize; i++)
+        texdata[i] = rand();
+
+      for (i = 4; i < bytesize; i += 8)
+        *(u32 *)&texdata[i] = 0x1B1B1B1B;
```

Comment:
Yeah, this was for testing, it makes the colors vertical, but it's a pattern of four vertical bars, one for each different color in the block - it was good for development, but I'll remove it because it'd be better a better test to randomize the selectors too.